### ISSUE TITLE
[improvement] Add an `optional()` method for prompts triggering the return of `Option<T>`

### ISSUE THREAD

**Issue by cylewitruk** [fadeevab/cliclack#68]
Today we've got `required(bool)` which steers whether or not input is required, however this triggers the return of a default value when the user doesn't provide any input. I find that I'm doing a bit of str.is_empty() for truly optional inputs, which doesn't feel so clean. It would be more idiomatic to make use of `Option<T>` when a field is truly optional.

I suggest adding an `optional()` modifier method which changes the return type to `Option<T>` and `None` is returned upon lack of input (and no `default_input()` is specified). Suggesting a new method so as not to break backwards-compatibility with usages of `required()` (which could be deprecated after the introduction of `optional()`).

Edit: This might be able to be done implicitly as well using `Any` and downcasting/checking if the inferred return type is an `Option<>`, but that seems like more work :)

**Comment by fadeevab** [comment]
@cylewitruk Do you mean that you need `required(false)` with an empty string converted into `None`?

**Comment by cylewitruk** [comment]
> @cylewitruk Do you mean that you need `required(false)` with an empty string converted into `None`?

Yeah, like:
```rust
let input: Option<String> = input("...").required(false).interact()?;```
But that doesn't work today, hence why I suggested a new method to transpose the type to an `Option<T>`. Maybe there's a way to achieve this today that I'm not seeing?

**Comment by fadeevab** [comment]
@cylewitruk We can't overcome the stale API `interact<T: FromStr>() -> Result<T>`. Basically, we can't implement `FromStr` for `Option<String>`: both the trait and the type are external to the crate. It's not possible to return `Option<String>` implicitly, there is gonna be an error.

I could suggest implementing your own `StrOption` and implement `FromStr` for `StrOption`, which basically fits your custom understanding of what is optional ("empty string is None"). You can even further convert `StrOption` to `Option`.

```rust
use std::str::FromStr;

enum StrOption {
    None,
    Some(String),
}

impl FromStr for StrOption {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            Ok(StrOption::None)
        } else {
            Ok(StrOption::Some(s.to_string()))
        }
    }
}

impl From<StrOption> for Option<String> {
    fn from(opt: StrOption) -> Option<String> {
        match opt {
            StrOption::None => None,
            StrOption::Some(s) => Some(s),
        }
    }
}

fn main() {
    let a: &str = "";
    let opt: Option<String> = a.to_string().parse::<StrOption>().unwrap().into();
    println!("{:?}", opt);
}
```

**Comment by fadeevab** [comment]
@cylewitruk With custom `StrOption` the following snipper will work:

```rust
let opt: StrOption = input().interact()?;
let opt2: Option = input().interact::<StrOption>()?.into();
```

**Comment by fadeevab** [comment]
I've included an example of optional input in the `examples/optional.rs`: https://github.com/fadeevab/cliclack/commit/bfb67d9e8d86e55b75f5248254db6445c93b2572

@cylewitruk I closed PR #72 with this issue 🤧 Again, unless a brighter idea appears.
